From da9c5c7e7536101d6f86c35e62fe32edbf18b597 Mon Sep 17 00:00:00 2001
From: liushuyu <liushuyu011@gmail.com>
Date: Sat, 4 Apr 2026 16:13:04 +0800
Subject: [PATCH 2/4] FROMGIT: Change Mtime field to INT64 to adapt to new GRUB
 changes

* GRUB upstream commit 81f1962393f4403e2b6b127f23524a962a236afb changed
  mtime to `grub_int64_t`

Link: https://github.com/pbatard/EfiFs/pull/52
Signed-off-by: Kaiyang Wu <origincode@aosc.io>
---
 src/driver.h | 6 +++---
 src/file.c   | 1 -
 src/grub.c   | 6 +++---
 3 files changed, 6 insertions(+), 7 deletions(-)

diff --git a/src/driver.h b/src/driver.h
index 93f323f..47faf02 100644
--- a/src/driver.h
+++ b/src/driver.h
@@ -161,7 +161,7 @@ typedef struct _EFI_GRUB_FILE {
 	EFI_FILE               EfiFile;
 	BOOLEAN                IsDir;
 	INT64                  DirIndex;
-	INT32                  Mtime;
+	INT64                  Mtime;
 	CHAR8                 *path;
 	CHAR8                 *basename;
 	INTN                   RefCount;
@@ -191,7 +191,7 @@ typedef struct _GRUB_DIRHOOK_INFO {
 	UINT32                 MtimeSet:1;
 	UINT32                 CaseInsensitive:1;
 	UINT32                 InodeSet:1;
-	INT32                  Mtime;
+	INT64                  Mtime;
 	UINT64                 Inode;
 } GRUB_DIRHOOK_INFO;
 
@@ -217,7 +217,7 @@ extern CHAR16 *GrubGetUuid(EFI_FS *This);
 extern BOOLEAN GrubFSProbe(EFI_FS *This);
 extern EFI_STATUS GrubDeviceInit(EFI_FS *This);
 extern EFI_STATUS GrubDeviceExit(EFI_FS *This);
-extern VOID GrubTimeToEfiTime(const INT32 t, EFI_TIME *tp);
+extern VOID GrubTimeToEfiTime(const INT64 t, EFI_TIME *tp);
 extern VOID CopyPathRelative(CHAR8 *dest, CHAR8 *src, INTN len);
 extern EFI_STATUS GrubOpen(EFI_GRUB_FILE *File);
 extern EFI_STATUS GrubDir(EFI_GRUB_FILE *File, const CHAR8 *path,
diff --git a/src/file.c b/src/file.c
index fcb79fc..41866c1 100644
--- a/src/file.c
+++ b/src/file.c
@@ -298,7 +298,6 @@ DirHook(const CHAR8 *name, const GRUB_DIRHOOK_INFO *DirInfo, VOID *Data)
 		return (INT32) Status;
 	}
 
-	// Oh, and of course GRUB uses a 32 bit signed mtime value (seriously, wtf guys?!?)
 	if (DirInfo->MtimeSet)
 		GrubTimeToEfiTime(DirInfo->Mtime, &Time);
 	CopyMem(&Info->CreateTime, &Time, sizeof(Time));
diff --git a/src/grub.c b/src/grub.c
index a7feaa4..eaac21f 100644
--- a/src/grub.c
+++ b/src/grub.c
@@ -263,9 +263,9 @@ static const unsigned short int __mon_yday[2][13] = {
 
 /* Compute an EFI_TIME representation of a GRUB's mtime_t */
 VOID
-GrubTimeToEfiTime(const INT32 t, EFI_TIME *tp)
+GrubTimeToEfiTime(const INT64 t, EFI_TIME *tp)
 {
-	INT32 days, rem, y;
+	INT64 days, rem, y;
 	const unsigned short int *ip;
 
 	days = t / SECS_PER_DAY;
@@ -286,7 +286,7 @@ GrubTimeToEfiTime(const INT32 t, EFI_TIME *tp)
 
 	while (days < 0 || days >= (__isleap (y) ? 366 : 365)) {
 		/* Guess a corrected year, assuming 365 days per year. */
-		INT32 yg = y + days / 365 - (days % 365 < 0);
+		CONST INT64 yg = y + days / 365 - (days % 365 < 0);
 
 		/* Adjust DAYS and Y to match the guessed year. */
 		days -= ((yg - y) * 365
-- 
2.55.0

